ResolvedTS/ResolvedTS.tla

```diff
@@ -0,0 +1,240 @@
+
+Max(s) ==
+    IF s = {} THEN 0
+    ELSE CHOOSE n \in s: \A m \in s: n \geq m
+
+FetchTS ==
+    /\ nextTS' = nextTS + 1 \* Try not to blew up
+                            \* state space by increasing nextTS
+    /\ LET resolvedTS == IF Len(resolvedQueue) = 0 THEN 0
+                         ELSE resolvedQueue[Len(resolvedQueue)].ts
+           cTS == nextTS - 0 \* A random offset, means cTS may be stale.
+       IN /\ chan' = Append(chan, [type |-> "FetchTS", ts |-> cTS])
+          /\ UNCHANGED <<rmState, reState, chanOffset,
+                         resolvedQueue, outputTSQueue>>
+
+TryResolve(cTS) ==
+    LET nonZero == {r \in RM: reState[r].minPTS > 0}
+        min == Min({reState[re].minPTS: re \in nonZero} \cup {cTS})
+    IN  LET nextIndex == Len(outputTSQueue) + 1
+            lastResolvedTS ==   IF resolvedQueue = << >> THEN 0
+                                ELSE resolvedQueue[Len(resolvedQueue)].ts
+            newResolvedTS == Max({lastResolvedTS, min})
+        IN  IF newResolvedTS # lastResolvedTS
+            THEN /\ outputTSQueue' = Append(outputTSQueue, newResolvedTS)
+                 /\ resolvedQueue' = Append(resolvedQueue,
+                                                [ts |-> newResolvedTS,
+                                                index |-> nextIndex + 1])
+            ELSE UNCHANGED <<resolvedQueue, outputTSQueue>>
+
+\* Check if the type of the next message in the chan equals to ty.
+CanHandleMsg(ty) ==
+    /\ chanOffset < Len(chan)
+    /\ chanOffset' = chanOffset + 1
+    /\ chan[chanOffset + 1].type = ty
+
+HandleMsgFetchTS ==
+    /\ CanHandleMsg("FetchTS")
+    /\  LET m == chan[chanOffset + 1]
+        IN  /\ TryResolve(m.ts)
+            /\ UNCHANGED <<nextTS, rmState, reState, chan>>
+
+HandleMsgPrewrite ==
+    /\ CanHandleMsg("Prewrite")
+    /\  LET m == chan[chanOffset + 1]
+        IN  /\ reState' = [reState EXCEPT ![m.rm].minPTS = m.ts]
+            /\ UNCHANGED <<nextTS, rmState, outputTSQueue, resolvedQueue, chan>>
+
+HandleMsgCommit ==
+    /\ CanHandleMsg("Commit")
+    /\  LET m == chan[chanOffset + 1]
+        IN  /\ reState' = [reState EXCEPT ![m.rm].minPTS = 0]
+            /\ outputTSQueue' = Append(outputTSQueue, m.ts)
+            /\ UNCHANGED <<nextTS, rmState, resolvedQueue, chan>>
+
+HandleMsgRollback ==
+    /\ CanHandleMsg("Rollback")
+    /\  LET m == chan[chanOffset + 1]
+        IN  /\ reState' = [reState EXCEPT ![m.rm].minPTS = 0]
+            /\ UNCHANGED <<nextTS, rmState, resolvedQueue, chan>>
+
+HandleMsg ==
+    \/ HandleMsgPrewrite
+    \/ HandleMsgCommit
+    \/ HandleMsgFetchTS
+
+RTNext ==
+    \/ TxnEvents
+    \/ FetchTS
+    \/ HandleMsg
+
+RTNextWithPessimistic ==
+    \/ TxnEventsWithPessimistic
+    \/ FetchTS
+    \/ HandleMsg
+
+RTSpec == RTInit /\ [][RTNext]_vars
+
+RTSpecWithPessimistic == RTInit /\ [][RTNextWithPessimistic]_vars
+
+\* Even if there is no write, it can always output
+\* resolved ts solely by FetchTS.
+ResolveLiveness ==
```

Comment:
I think this variant is always true, and the feature you  really want to verify is that if there are any writes at some point, the `resolvedTS` would still forward,  which also means at least one state has `resolvedTS` == `fetchTS`